Remove oesign's dependency on libsgx_enclave_common and libsgx_dcap_ql

nonpolarity opened the pull request:
A static library liboemr.a is introduced, which has no dependency on these shared libraries.
All the dependent sentences are stripped away through a macro "OEMR" during preprocessing.

Fixes #2544 
Fixes #2264

nonpolarity commented:
bors try

nonpolarity commented:
bors try

nonpolarity commented:
bors try

nonpolarity commented:
bors try

nonpolarity commented:
bors try

nonpolarity commented:
bors try

nonpolarity commented:
bors try

nonpolarity commented:
bors try

radhikaj commented on host/CMakeLists.txt:
```diff
@@ -488,3 +488,321 @@ install(TARGETS oehostapp EXPORT openenclave-targets)
+if (UNIX)
+	set(PLATFORM_OEMR_SRC
+		#    ../common/cert.c
+		#    crypto/openssl/asn1.c
```
Why are these commented out?

anakrish replied:
I agree with @radhikaj. These commented lines could be removed.

nonpolarity replied:
Thanks, I will clean up when everything is done.

anakrish commented:
@nonpolarity @radhikaj @CodeMonkeyLeet 

My understanding was that we were planning to solve this via delayed loading of the intel shared libraries as opposed to the current static linking of the intel shared libraries.

Though this PR seems to address the specific issue of design, there are other related issues that a delayed loading approach would solve:

By delayed loading,
- oesign will not depend on intel sgx libraries
- Purely simulation workflow will not depend of sgx libraries being present in the system
- Avoids FLC vs NON FLC issues where the published OE SDK only works on SGX FLC machines

Here is an example of how we already delay load azure quote provider library:
https://github.com/openenclave/openenclave/blob/11090df10b820755eb710b08a473f96d5edb3282/host/sgx/windows/sgxquoteproviderloader.c

Here is another example of how we delay load oedebugrt.dll on Windows:
https://github.com/openenclave/openenclave/blob/11090df10b820755eb710b08a473f96d5edb3282/host/sgx/windows/debugrtbridge.c

radhikaj commented:
> @nonpolarity @radhikaj @CodeMonkeyLeet
> 
> My understanding was that we were planning to solve this via delayed loading of the intel shared libraries as opposed to the current static linking of the intel shared libraries.
> 
> Though this PR seems to address the specific issue of design, there are other related issues that a delayed loading approach would solve:
> 
> By delayed loading,
> 
> * oesign will not depend on intel sgx libraries
> * Purely simulation workflow will not depend of sgx libraries being present in the system
> * Avoids FLC vs NON FLC issues where the published OE SDK only works on SGX FLC machines
> 
> Here is an example of how we already delay load azure quote provider library:
> https://github.com/openenclave/openenclave/blob/11090df10b820755eb710b08a473f96d5edb3282/host/sgx/windows/sgxquoteproviderloader.c
> 
> Here is another example of how we delay load oedebugrt.dll on Windows:
> https://github.com/openenclave/openenclave/blob/11090df10b820755eb710b08a473f96d5edb3282/host/sgx/windows/debugrtbridge.c

@anakrish When we started off, we were trying to address these issues: https://github.com/openenclave/openenclave/issues/2264 and https://github.com/openenclave/openenclave/issues/2544

@nonpolarity has addressed them. We can take further action to remove dependencies on the host side completely, but can do that after.

anakrish commented on host/CMakeLists.txt:
```diff
@@ -488,3 +488,321 @@ install(TARGETS oehostapp EXPORT openenclave-targets)
+if (OE_SGX AND WIN32)
+  # oedebugrt is accessed via a bridge on Win32 and need not be linked.
+  list(APPEND PLATFORM_OEMR_SRC
+    sgx/windows/debugrtbridge.c)
```
This seems off. oesign is not concerned with debugging and would not need to do include debugrt stuff.

nonpolarity replied:
updated, thanks.

anakrish commented on host/CMakeLists.txt:
```diff
@@ -488,3 +488,321 @@ install(TARGETS oehostapp EXPORT openenclave-targets)
+#add_dependencies(oemr syscall_untrusted_edl)
+#add_dependencies(oemr tee_untrusted_edl)
+if(OE_SGX)
+  add_dependencies(oemr sgx_untrusted_edl switchless_untrusted_edl)
```
oesign most likely does not need any edl stuff either.

nonpolarity replied:
Since sgx_untrusted_edl  depends on sgx_u.h, add_dependencies can ensure dependency on sgx_u.h. I will try to find if there is a straightforward way for dependency on file.

anakrish commented on host/CMakeLists.txt:
```diff
@@ -488,3 +488,321 @@ install(TARGETS oehostapp EXPORT openenclave-targets)
+endif ()
+
+# Install targets
+install(TARGETS oemr EXPORT openenclave-targets
```
Since oems is used only to build oesign, and is not intended to be used by OE SDK user, we need not install it.

nonpolarity replied:
Good point, this is just a internal and intermediate library.

nonpolarity commented:
bors try

nonpolarity commented:
bors try

anakrish commented on host/CMakeLists.txt:
```diff
@@ -307,7 +307,7 @@ endif()
 add_dependencies(oehost tee_untrusted_edl)
 if(OE_SGX)
-  add_dependencies(oehost sgx_untrusted_edl switchless_untrusted_edl)
+  add_dependencies(oehost sgx_untrusted_edl)
```
Accidental removal of switchless_untrusted_edl?

nonpolarity replied:
Really bad accident, thanks.

radhikaj commented:
@anakrish Please discuss various approaches you have for this including delay loading and stubbing and removing host side dependencies.

nonpolarity commented:
bors try

nonpolarity commented:
> 
> 
> @nonpolarity Thanks for addressing the comments. Looks much simpler to understand.
> 
> Couple of suggestions:
> 
>     * It is not clear what `oemr` means. It would be good to use more descriptive names for the library as well as for the C preprocessor define
> 
>     * Since oemr and oehost do not seem to be related, it would be cleaner to create an oemr folder with just a CMakeLists.txt and a README.md. oemr/CMakeLists.txt could refer to files in host, and the README.md could describe the intent of the library.

Thanks for these suggestions. The library name is changed to liboehostmr, following naming convention of other host side libraries. The cmake config file is also moved to a subdir.

anakrish commented on host/OEHOSTMR/CMakeLists.txt:
```diff
@@ -0,0 +1,189 @@
+    crypto/openssl/key.c
+    crypto/openssl/rsa.c
+    crypto/openssl/sha.c
+	linux/hostthread.c
```
Kindly fix the spacing issues in this file.

nonpolarity replied:
Thanks, tab is replace with space.

anakrish commented on host/CMakeLists.txt:
```diff
@@ -488,3 +488,5 @@ install(TARGETS oehostapp EXPORT openenclave-targets)
   ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}/openenclave/host
   COMPONENT OEHOSTVERIFY)
+
+include(OEHOSTMR/CMakeLists.txt)
```
It would make sense to add a line `add_subdirectory(oehostmr)` here at the top-level CMakeLists.txt:
https://github.com/openenclave/openenclave/blob/99410b10b4782d33e76819bc681e651e96cd9739/CMakeLists.txt#L213

nonpolarity replied:
@anakrish since oehostmr is part of host as oehostverify, is it better to stay under host directory? Actually the working directory for oehostmr is also host not host/oehostmr, because change working directory requires lots of change of cmake config and even head file reference. So right now include(OEHOSTMR/CMakeLists.txt) is a easy way to separate the config file. Let me know if I have some misunderstanding.

anakrish commented:
@CodeMonkeyLeet @nonpolarity 
>
> Thanks for tackling this! Since this is still in draft, have you considered making oehostmr a variant of the oehost build just as oehostverify is generated from the same CMakeLists.txt today?
>

Responding since I had asked Weichen to move the code out from host/CMakeLists.txt into a separate oehostmr/CMakeLists.txt. I prefer separate cmakefile since there is no change to oehost component other than conditional compilation in some source files. host/CMakeLists.txt is already a bit too unwieldy. This way, If it turns out that delay loading would solve the same problem, then it is easier to remove oehostmr component entirely; otherwise we may have to revert oehost's CMakeLists.txt to an earlier version.

dthaler commented on host/OEHOSTMR/CMakeLists.txt:
```diff
@@ -0,0 +1,189 @@
+    windows/syscall.c
+    windows/time.c)
+else()
+  message(FATAL_ERROR "Unknown OS. Only supported OSes are Linux and Windows")
```
grammar: "Only" -> "The only"

nonpolarity replied:
Fixed, thanks.

dthaler commented on host/OEHOSTMR/CMakeLists.txt:
```diff
@@ -0,0 +1,189 @@
+
+  set(PLATFORM_FLAGS "-m64")
+elseif(OE_TRUSTZONE)
+    list(APPEND PLATFORM_HOST_MR_SRC
```
nit: indentation seems off

nonpolarity replied:
Tab is fixed, thanks.

dthaler commented on host/OEHOSTMR/CMakeLists.txt:
```diff
@@ -0,0 +1,189 @@
+  endif()
+
+  set(PLATFORM_FLAGS "-m64")
+elseif(OE_TRUSTZONE)
```
nit: use consistent spacing.  Usually this file has a space before (like in line 13 etc) but here doesn't.  Similarly line 97 and 110 have different styles. Pick a style and be consistent throughout this file.

nonpolarity replied:
Fixed, thanks.

dthaler commented on host/OEHOSTMR/README.md:
```diff
@@ -0,0 +1,3 @@
+The `liboehostmr` is a subset of liboehost, and it's main functionality is to get the enclave's measurement. Following other host libraries' naming convention, this library starts with liboehost, and `mr` is short for `measurement report`.
```
grammar: "it's" should be "its"

nonpolarity replied:
Fixed, thanks.

dthaler commented on host/OEHOSTMR/README.md:
```diff
@@ -0,0 +1,3 @@
+The `liboehostmr` is a subset of liboehost, and it's main functionality is to get the enclave's measurement. Following other host libraries' naming convention, this library starts with liboehost, and `mr` is short for `measurement report`.
+
+To reuse source code, a macro `OEHOSTMR` is used that some sentences refering to libsgx_enclave_common are stripped off. Thus this library has no dependency on libsgx_enclave_common(also libsgx_dcap_ql).
```
grammar: can't parse "is used that some sentences … are stripped off".
Please reword.   Also add space before open paren.

nonpolarity replied:
Fixed, thanks.

CodeMonkeyLeet commented:
> @anakrish 
> 
> > Thanks for tackling this! Since this is still in draft, have you considered making oehostmr a variant of the oehost build just as oehostverify is generated from the same CMakeLists.txt today?
> 
> Responding since I had asked Weichen to move the code out from host/CMakeLists.txt into a separate oehostmr/CMakeLists.txt. I prefer separate cmakefile since there is no change to oehost component other than conditional compilation in some source files. host/CMakeLists.txt is already a bit too unwieldy. This way, If it turns out that delay loading would solve the same problem, then it is easier to remove oehostmr component entirely; otherwise we may have to revert oehost's CMakeLists.txt to an earlier version.

@nonpolarity Please feel free to disregard this comment then; I don't feel strongly about this and I didn't mean to randomize you.

nonpolarity commented on host/OEHOSTMR/CMakeLists.txt:
```diff
@@ -0,0 +1,189 @@
+# Copyright (c) Open Enclave SDK contributors.
```
Yes there is a dilemma that the dir is not a  sub-dir of source code as sgx and crypto, neither a dependent cmake sub-module of host. Thus this dir is in upper case. I will change it to host/measure.

nonpolarity commented on host/sgx/sgxload.c:
```diff
@@ -418,12 +423,14 @@ oe_result_t oe_sgx_create_enclave(
      * mode or on Linux Kabylake machines. */
     if (context->type == OE_SGX_LOAD_TYPE_CREATE)
     {
+#if !defined(OEHOSTMR)
```
Good suggestion, a return code OE_NOT_IMPLEMENTED is added.

CodeMonkeyLeet commented on host/OEHOSTMR/README.md:
```diff
@@ -0,0 +1,3 @@
+The `liboehostmr` is a subset of liboehost, and it's main functionality is to get the enclave's measurement. Following other host libraries' naming convention, this library starts with liboehost, and `mr` is short for `measurement report`.
+
+To reuse source code, a macro `OEHOSTMR` is used that some sentences refering to libsgx_enclave_common are stripped off. Thus this library has no dependency on libsgx_enclave_common(also libsgx_dcap_ql).
```
```suggestion
The `liboehostmr` library reuses the source code from `liboehost` by using the preprocessor definition `OEHOSTMR`. When specified, `OEHOSTMR` removes any code references to libsgx_enclave_common or libsgx_dcap_ql so that `liboehostmr` has no dependencies on those libraries.
```

CodeMonkeyLeet replied:
It looks like there's a merge conflict and both upper case and lower case versions of the file exist in this PR now?

nonpolarity replied:
Thanks for rewriting the whole paragraph! Yes I forget to use `git mv` but `mv` only, now it is fixed.

nonpolarity commented on common/result.c:
```diff
@@ -130,6 +130,8 @@ const char* oe_result_str(oe_result_t result)
             return "OE_ALREADY_EXISTS";
         case OE_ALREADY_INITIALIZED:
             return "OE_ALREADY_INITIALIZED";
+        case OE_NOT_IMPLEMENTED:
```
Got it, thanks.

nonpolarity commented:
bors try

nonpolarity commented:
bors try

nonpolarity commented:
bors try

radhikaj commented:
bors r+

radhikaj commented:
@nonpolarity Thanks so much or this fix.

jhand2 commented:
Awesome job getting this done! People have been complaining about this in gh issues for a while
